### PR TITLE
fix(install): support Windows PowerShell 5.1 and non-ASCII install paths in cmd wrapper (#483)

### LINKED ISSUE #483: packaging: install.ps1 broken on stock Windows PowerShell 5.1 (3-arg Join-Path); cmd wrapper written as ASCII
## Summary

The Windows installer fails its two main audiences: the documented `irm | iex` one-liner dies with a parameter
binding error on the OS-default Windows PowerShell 5.1 (every stock Windows 10/11 box), and even where it runs, the
generated `envdrift.cmd` wrapper is corrupted for any user profile containing non-ASCII characters. CI cannot catch
either: the installer e2e job forces `shell: pwsh` (PowerShell 7) on runners with ASCII profile paths. Every item
below was reproduced twice (original reviewer + independent verifier) on origin/main commit cb68c97.

## Findings

- [ ] **install.ps1 uses 3-argument `Join-Path` (PowerShell 6+ only): the documented `irm | iex` install dies on
  stock Windows PowerShell 5.1 before creating anything.** `install.ps1:223`
  **Repro:** in the built-in "Windows PowerShell" (5.1) console:
  `irm https://raw.githubusercontent.com/jainal09/envdrift/main/install.ps1 | iex`; minimal probe:
  `Join-Path C:\x "Scripts" "python.exe"` → binding error in 5.1, succeeds in pwsh 7.
  **Expected:** the documented one-liner works in the OS-default shell (nest two-arg calls, the pattern already used
  at install.ps1:43-45), or README/docs explicitly require pwsh.
  **Actual:** "A positional parameter cannot be found that accepts argument"; with `$ErrorActionPreference='Stop'`
  (:30) the script aborts inside Initialize-Venv. Only PowerShell 7+ users can install.
  **Verifier note:** MS Learn confirms 5.1 Join-Path has no `-AdditionalChildPath`; the same form recurs at :238 and
  :267; installer-e2e.yml:74 forces `shell: pwsh`, so CI can never catch this.

- [ ] **install.ps1 writes the `envdrift.cmd` wrapper with `-Encoding ASCII` (the comment claims UTF8): non-ASCII
  Windows profile paths are mangled to `?`, leaving a dead wrapper.** `install.ps1:274`
  **Repro:** run install.ps1 with a non-ASCII `$HOME` (e.g. `C:\Users\José`); `type $HOME\.envdrift\bin\envdrift.cmd`
  shows `C:\Users\Jos?\...`. Deterministic: `[System.Text.Encoding]::ASCII.GetBytes('José')` → `Jos?`.
  **Expected:** the wrapper preserves the real path (unicode-safe encoding consistent with the comment's intent,
  like the .ps1 wrapper written UTF8 right below at :278-283).
  **Actual:** the primary CMD entry point — the exact command Show-PathInstructions tells the user to run (:451) —
  fails with "The system cannot find the path specified" while the installer reports success.
  **Verifier note:** plain UTF8-with-BOM can also break a .cmd first line under PS 5.1; the ANSI/OEM codepage is the
  safe target for the cmd wrapper.

## Suggested fix direction

- Replace the three 3-argument calls (install.ps1:223, :238, :267) with nested two-arg form:
  `Join-Path (Join-Path $VenvDir 'Scripts') 'python.exe'`.
- Write the cmd wrapper in the system ANSI/OEM codepage (or another encoding verified to round-trip non-ASCII paths
  in cmd.exe), and fix the misleading comment.
- Add an installer-e2e matrix leg using `shell: powershell` (5.1) so the OS-default shell is exercised in CI.

_Part of the 2026-06 final deep-review (batch: CI-2). Each item was independently reproduced and re-verified before filing._

### PR DESCRIPTION
**Pull request by jainal09**
## Summary

`install.ps1` failed both of its main audiences (#483): the documented `irm | iex` one-liner died
with a parameter binding error on the OS-default Windows PowerShell 5.1, and the generated
`envdrift.cmd` wrapper was dead for any user profile containing non-ASCII characters. CI forced
`shell: pwsh`, so neither could ever be caught. This PR ships the regression tests first (proven
failing against unfixed code), then the fix, plus a PowerShell 5.1 CI leg so it cannot regress.

## Fixes

| Issue checklist item | Change |
| --- | --- |
| 3-argument `Join-Path` (PS 6+ only) at `install.ps1:223/:238/:267` kills `irm \| iex` on stock 5.1 | Nested two-arg calls: `Join-Path (Join-Path $VenvDir "Scripts") <exe>` at all three sites |
| `envdrift.cmd` written `-Encoding ASCII` (comment claimed UTF8) mangles non-ASCII profile paths to `?`, leaving a dead wrapper | Wrapper now resolves the venv at runtime via `%~dp0..\venv\Scripts\envdrift.exe` — the file stays pure ASCII, so it round-trips for *any* profile path under cmd.exe's OEM codepage (the issue's verifier note: UTF8-with-BOM breaks the `.cmd` first line under 5.1, so writing the absolute path in UTF-8 was not a safe fix). Misleading comment replaced with an accurate one |
| `installer-e2e.yml` forces `shell: pwsh`, so CI can never catch 5.1 breakage | Windows job matrix now runs `shell: [pwsh, powershell]` (powershell = the genuine Windows PowerShell 5.1 engine). Also added a `Verify CMD wrapper` step that actually executes `envdrift.cmd` via `cmd.exe` — previously a dead cmd wrapper passed CI because only the `.ps1` wrapper was exercised |

Additional hardening required to make the install actually complete under 5.1 / stay 5.1-safe:

- The silent pip upgrade used `2>$null` under `$ErrorActionPreference='Stop'`; in Windows
  PowerShell 5.1 any native stderr line then becomes a terminating `NativeCommandError`
  (fixed only in PS 7.2+). The preference is relaxed around that single best-effort call.
- `docs/getting-started/installation.md` now states the installer works on both Windows
  PowerShell 5.1 and `pwsh`, with non-ASCII install paths.
- Filed #501 for a pre-existing, out-of-scope bug found during this work: `& @script:PythonCmd`
  splats the array into a single command name ("py -3" is not recognized), breaking
  py-launcher-only installs.

## Test evidence

New file `tests/unit/test_install_ps1.py` drives the **real** installer code under a **real**
PowerShell engine (AST-extracted, verbatim — no re-implementation), skip-gated on PowerShell
availability (preinstalled on all GitHub runners):

- `test_join_path_calls_are_windows_powershell_51_compatible` — every `Join-Path` must bind under
  the 5.1 signature (no 3-arg form / `-AdditionalChildPath`)
- `test_join_path_calls_bind_under_real_windows_powershell_51` — evaluates every `Join-Path` call
  under the genuine 5.1 binder (runs on the Windows CI lanes)
- `test_cmd_wrapper_survives_non_ascii_install_path` — runs the real `New-Wrappers` against a
  `José Müller` install dir and asserts the wrapper resolves to the venv exe, starts with
  `@echo off` (no BOM), and stays pure ASCII
- `test_install_ps1_is_ascii_or_carries_a_bom`, `test_install_ps1_parses_without_errors` — guards
  (5.1 reads BOM-less files as ANSI)

**Proven failing pre-fix** (against `origin/main`, pwsh 7.5.2):

```text
FAILED test_join_path_calls_are_windows_powershell_51_compatible
  AssertionError: install.ps1 contains Join-Path calls that do not bind under Windows
  PowerShell 5.1 (nest two-argument calls instead): [{'Line': 223, 'Text': 'Join-Path $VenvDir
  "Scripts" "python.exe"'}, {'Line': 238, ...}, {'Line': 267, ...}]
FAILED test_cmd_wrapper_survives_non_ascii_install_path
  AssertionError: envdrift.cmd does not resolve to the venv executable for a non-ASCII install
  path: wrapper targets '.../Jos? M?ller/.envdrift/venv/Scripts/envdrift.exe',
  expected '.../José Müller/.envdrift/venv/Scripts/envdrift.exe'
```

**Post-fix:** `4 passed, 1 skipped` locally (the 5.1-binder test runs on Windows CI); full suite
`uv run pytest` green; `ruff check` / `ruff format --check` / `pyrefly` / `bandit` /
`make lint-docs` / `actionlint` all clean.

Fixes #483

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<a href="https://cubic.dev/pr/jainal09/envdrift/pull/503?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

<h3>Greptile Summary</h3>

This PR fixes two documented installation failures on Windows PowerShell 5.1 (`irm | iex`) and adds a PowerShell 5.1 CI lane so regressions can't sneak back in. The root causes were a 3-argument `Join-Path` (PS 6+ only) that killed the installer on stock Windows boxes, and a CMD wrapper that baked in the absolute venv path with ASCII encoding, mangling non-ASCII profile paths.

- **`install.ps1`**: Three `Join-Path a b c` calls replaced with nested two-argument form; CMD wrapper now uses `%~dp0..\\venv\\Scripts\\envdrift.exe` (runtime-relative, pure ASCII); `$ErrorActionPreference` is now saved and restored around the best-effort pip upgrade.
- **`.github/workflows/installer-e2e.yml`**: Windows job matrix gains a `shell: [pwsh, powershell]` axis so genuine 5.1 runs in CI; a new \"Verify CMD wrapper\" step exercises `envdrift.cmd` via `cmd.exe`.
- **`tests/unit/test_install_ps1.py`**: New file with 6 regression tests driving the real installer code under a real PowerShell engine (AST-extracted verbatim); proven failing against the unfixed code and green post-fix.

<h3>Confidence Score: 5/5</h3>

Safe to merge — all three bugs are precisely targeted and verified against the real PowerShell 5.1 engine; the regression test suite proves each fix on both the pre- and post-fix codebase.

The changes are surgical: three one-line Join-Path rewrites, a CMD wrapper that trades an embedded absolute path for a runtime-relative one, and an EAP save/restore block. Each fix is backed by a dedicated regression test driven against the real PowerShell AST binder, and the CI matrix now exercises the genuine Windows PowerShell 5.1 engine end-to-end. No logic was removed or restructured; the installer's overall flow is unchanged.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| install.ps1 | Three 3-arg Join-Path calls replaced with nested 2-arg calls; CMD wrapper switched from embedded absolute path to runtime-relative %~dp0 reference; EAP save/restore added around best-effort pip upgrade — all fixes are correct and targeted. |
| .github/workflows/installer-e2e.yml | Windows matrix gains shell: [pwsh, powershell] axis for genuine 5.1 coverage; new Verify CMD wrapper step exercises envdrift.cmd via cmd.exe; all inline Join-Path calls use nested 2-arg form for 5.1 compatibility. |
| tests/unit/test_install_ps1.py | New regression test file with 6 tests using real PowerShell engine AST extraction; harness files use utf-8-sig encoding, subprocess gets -ExecutionPolicy Bypass, and output decoded with errors=replace — all edge cases for 5.1 compatibility addressed. |
| docs/getting-started/installation.md | Three-line addition clarifying that the installer works on both Windows PowerShell 5.1 and pwsh, and supports non-ASCII install paths. |

<sub>Reviews (3): Last reviewed commit: ["Merge branch &#39;main&#39; into fix/483-install..."](https://github.com/jainal09/envdrift/commit/8591735b532bf401729d54e93867fd4b7f7f980a) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=37198219)</sub>

### CODE REVIEW

File: install.ps1 [hunk 1]
```diff
@@ -235,10 +237,15 @@ function Initialize-Venv {
...
+    $script:VenvPip = Join-Path (Join-Path $VenvDir "Scripts") "pip.exe"
 
-    # Upgrade pip silently
+    # Upgrade pip silently (best-effort). Keep $ErrorActionPreference relaxed
+    # around the redirection: under Windows PowerShell 5.1 with 'Stop', any
+    # native stderr line combined with 2>$null becomes a terminating
+    # NativeCommandError (fixed only in PowerShell 7.2+).
+    $ErrorActionPreference = "Continue"
     & $script:VenvPython -m pip install --upgrade pip 2>$null | Out-Null
+    $ErrorActionPreference = "Stop"
```
[the remark this answers is not in the document]
Comment (jainal09) [reply install.ps1]: Fixed in f02fec2: the entry `$ErrorActionPreference` is now saved and restored in a `try/finally` instead of hardcoding `"Stop"`.

One correction to the stated impact, verified on a real Windows PowerShell 5.1.26100 box: a *normal* call can never replace the caller's preference — a preference-variable assignment inside a function creates a function-local, so with the old code a `SilentlyContinue` caller still saw `SilentlyContinue` after `Initialize-Venv` returned. The hardcode only bites when the body runs in the caller's scope (dot-invocation: `. Initialize-Venv`), where the old code really did leave the session at `Stop`. The new regression test (`test_initialize_venv_restores_caller_error_action_preference`) dot-invokes the real function under the genuine 5.1 engine with a network-free `--without-pip` venv; it was verified to fail against the previous code (`EAP_AFTER=Stop`) and passes now.

File: tests/unit/test_install_ps1.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+    _windows_powershell_51() is None,
+    reason="Windows PowerShell 5.1 only exists on Windows",
+)
+
+
+def _run_harness(
+    exe: str, harness: str, tmp_path: Path, *args: str
+) -> subprocess.CompletedProcess[str]:
+    harness_file = tmp_path / "harness.ps1"
+    harness_file.write_text(harness, encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (jainal09) [reply tests/unit/test_install_ps1.py]: Fixed in f02fec2: harness files are now written with `encoding="utf-8-sig"` as suggested, so future non-ASCII harness content can't be silently decoded as ANSI by 5.1.

While battle-testing the suite on a real Windows PowerShell 5.1.26100 box (stock execution policy, OEM console codepage) two more environment gaps surfaced and are hardened in the same commit: `_run_harness` now passes process-scoped `-ExecutionPolicy Bypass` (stock boxes default to `Restricted`, which rejects `-File` outright — GitHub runners allow scripts, so CI never saw it), and subprocess output is decoded with `errors="replace"` because 5.1 echoes non-ASCII install paths (e.g. `José Müller`) in the OEM codepage, which is not valid UTF-8. All 6 tests pass on the real 5.1 engine, including real venv creation and `envdrift.cmd` execution from a `C:\...\ünïcödé tëst\.envdrift` install dir.
